In [0]:
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import OneHotEncoder, StandardScaler
import numpy as np
import pandas as pd


iris = load_iris()
X = iris['data']
y = iris['target']

In [0]:
from sklearn.preprocessing import LabelEncoder
encoder =  LabelEncoder()
y1 = encoder.fit_transform(y)
Y = pd.get_dummies(y1).values


In [3]:
X.shape

(150, 4)

In [4]:
Y.shape

(150, 3)

In [0]:
from sklearn.model_selection import train_test_split
X_train,X_test, y_train,y_test = train_test_split(X,Y,test_size=0.2,random_state=228) 

In [6]:
import keras
from keras.models import Sequential
from keras.layers import Dense, Activation

def build_model():
  model = Sequential([
      Dense(16, input_shape=(4,),activation='tanh'),
      Dense(32,activation='tanh'),
      Dense(3,activation='softmax'),
  ])
  model.compile(optimizer='adam',
              loss='categorical_crossentropy',
              metrics=['accuracy'])
  return model

Using TensorFlow backend.


In [7]:
model1 = build_model()
model1.summary()






Model: "sequential_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_1 (Dense)              (None, 16)                80        
_________________________________________________________________
dense_2 (Dense)              (None, 32)                544       
_________________________________________________________________
dense_3 (Dense)              (None, 3)                 99        
Total params: 723
Trainable params: 723
Non-trainable params: 0
_________________________________________________________________


In [8]:

model1.fit(X_train,y_train,epochs = 100)

Instructions for updating:
Use tf.where in 2.0, which has the same broadcast rule as np.where



Epoch 1/100





120/120 [==============================] - 1s 6ms/step - loss: 1.4601 - acc: 0.3500
Epoch 2/100
120/120 [==============================] - 0s 83us/step - loss: 1.3354 - acc: 0.3500
Epoch 3/100
120/120 [==============================] - 0s 80us/step - loss: 1.2191 - acc: 0.3500
Epoch 4/100
120/120 [==============================] - 0s 110us/step - loss: 1.1307 - acc: 0.3500
Epoch 5/100
120/120 [==============================] - 0s 97us/step - loss: 1.0617 - acc: 0.3750
Epoch 6/100
120/120 [==============================] - 0s 87us/step - loss: 1.0110 - acc: 0.6667
Epoch 7/100
120/120 [==============================] - 0s 94us/step - loss: 0.9766 - acc: 0.7000
Epoch 8/100
120/120 [==============================] - 0s 94us/step - loss: 0.9561 - acc: 0.7000
Epoch 9/100
120/120 [==============================] - 0s 91us/step - loss: 0.9251 - acc: 0.7000
Epoch 10/100
120/120 [===

In [9]:
y_pred = model1.predict(X_test)

y_test_class = np.argmax(y_test,axis=1)
y_pred_class = np.argmax(y_pred,axis=1)

print(y_test_class)
print(y_pred_class)

[0 0 1 1 1 1 1 0 0 1 1 2 1 1 2 0 2 2 1 0 1 1 0 0 1 2 2 1 2 2]
[0 0 1 1 1 1 1 0 0 1 1 2 1 1 2 0 2 2 1 0 1 1 0 0 1 2 2 1 2 2]


In [10]:
print(model1.get_weights())

[array([[ 0.13856044,  0.3324899 , -0.21610937,  0.04943543,  0.19390005,
        -0.06632356,  0.3554743 , -0.12060726, -0.13560684,  0.08320681,
        -0.2818772 ,  0.12545374,  0.04158588, -0.27185458,  0.21667124,
         0.11321702],
       [ 0.33275363, -0.02664353, -0.07806253, -0.35604656,  0.06843536,
        -0.26292774, -0.18677421, -0.16126104, -0.00232079,  0.3875494 ,
         0.07276726,  0.43924507, -0.20629838,  0.57715994, -0.63067955,
        -0.58064085],
       [-0.22417207, -0.1754693 ,  0.2680888 , -0.07854957,  0.519101  ,
        -0.10506185, -0.48912412,  0.05555359,  0.34044474, -0.40393162,
         0.17640814, -0.2604668 ,  0.21597   , -0.24550125,  0.17629659,
        -0.50271755],
       [-0.5925389 , -0.05064395, -0.34189084, -0.1520825 ,  0.40897936,
         0.6024598 ,  0.19742891,  0.65534985,  0.5925363 , -0.04908391,
        -0.2175926 , -0.6181363 ,  0.5516789 , -0.00600267,  0.1963614 ,
        -0.57839704]], dtype=float32), array([ 0.13742824

In [0]:
def update_NN(model): # add random vector to weights for mutation
  weights = model.get_weights()
  for i in range(len(weights)):
    weights[i] = np.random.uniform(-1,1,weights[i].shape)
  new_model = build_model()
  new_model.set_weights(weights)
  return new_model
  

In [12]:
from sklearn.metrics import accuracy_score

new_model = update_NN(model1)
y_pred = new_model.predict(X_test)

y_test_class = np.argmax(y_test,axis=1)
y_pred_class = np.argmax(y_pred,axis=1)

print(y_test_class)
print(y_pred_class)

[0 0 1 1 1 1 1 0 0 1 1 2 1 1 2 0 2 2 1 0 1 1 0 0 1 2 2 1 2 2]
[0 1 1 0 0 0 1 1 1 0 0 0 1 0 0 1 0 0 0 1 0 0 1 0 1 0 1 1 0 0]


In [0]:
def energy_fun(model):  # cost function using MSE
  y_pred = model.predict(X_train)
  mse = (np.square(y_pred - y_train)).mean(axis=None)
  return mse

In [14]:
energy_fun(new_model) # result of mutated model

0.22212753

In [15]:
energy_fun(model1) # result of backprop

0.019370927

In [0]:
def simulated_annealing(initial, generate_f, energy_f, T = 2 , alpha = 0.1, iters = 10, min_T = 0.05):
  best = initial
  e_best = energy_f(initial)
  x0 = initial
  while T > min_T:
    for i in range(iters):
      x = generate_f(x0)
      e_x = energy_f(x)
      e_x0 = energy_f(x0)
      if(e_x < e_x0):
        x0 = x
        if e_x < e_best:
          e_best = e_x
          best = x
      else:
        accept_ratio = np.exp(-e_x/T)/ np.exp(-e_x0/T)
        if accept_ratio < np.random.uniform(0.0,1.0):
          x0 = x
      print(T, i, e_best, e_x, e_x0)
    T = T * alpha
  return best,e_best















In [0]:
def accuracy(model):
  y_pred = model.predict(X_test)

  y_test_class = np.argmax(y_test,axis=1)
  y_pred_class = np.argmax(y_pred,axis=1)
  return accuracy_score(y_test_class, y_pred_class)



In [30]:
# run 1
initial = build_model()
T = 10.0
alpha = 0.80
T_min = 1
iters = 5

model,cost  = simulated_annealing(initial=initial,generate_f=update_NN, energy_f=energy_fun,T = 5)

cost

5 0 0.2764798 0.3273315 0.2764798
5 1 0.2764798 0.37411362 0.2764798
5 2 0.2764798 0.38740367 0.2764798
5 3 0.2764798 0.42791247 0.2764798
5 4 0.2764798 0.36527464 0.2764798
5 5 0.2764798 0.3648541 0.2764798
5 6 0.23668188 0.23668188 0.2764798
5 7 0.2239177 0.2239177 0.23668188
5 8 0.2239177 0.44255167 0.2239177
5 9 0.2239177 0.4136252 0.2239177
0.5 0 0.2239177 0.38258404 0.2239177
0.5 1 0.2239177 0.34018785 0.2239177
0.5 2 0.2239177 0.30889025 0.2239177
0.5 3 0.2239177 0.2635761 0.2239177
0.5 4 0.21505643 0.21505643 0.2239177
0.5 5 0.19535199 0.19535199 0.21505643
0.5 6 0.19535199 0.23984618 0.19535199
0.5 7 0.19535199 0.31433225 0.19535199
0.5 8 0.19535199 0.42101982 0.31433225
0.5 9 0.19535199 0.46287563 0.42101982


0.19535199

In [31]:
accuracy(model)

0.5

In [33]:
# run 2
initial = build_model()
T = 1000.0
alpha = 0.5
T_min = 10
iters = 5

model,cost  = simulated_annealing(initial=initial,generate_f=update_NN, energy_f=energy_fun,T = T, min_T = T_min, iters = iters, alpha = alpha)

cost

1000.0 0 0.26613054 0.4388989 0.26613054
1000.0 1 0.26613054 0.28311107 0.26613054
1000.0 2 0.26613054 0.44648573 0.26613054
1000.0 3 0.26613054 0.4550956 0.26613054
1000.0 4 0.26613054 0.43208963 0.26613054
500.0 0 0.26613054 0.41384202 0.26613054
500.0 1 0.15144482 0.15144482 0.26613054
500.0 2 0.15144482 0.40087196 0.15144482
500.0 3 0.15144482 0.34181944 0.15144482
500.0 4 0.15144482 0.41541016 0.15144482
250.0 0 0.15144482 0.34895557 0.15144482
250.0 1 0.15144482 0.36676848 0.15144482
250.0 2 0.15144482 0.3661289 0.15144482
250.0 3 0.15144482 0.31883895 0.15144482
250.0 4 0.15144482 0.41973814 0.15144482
125.0 0 0.15144482 0.45269626 0.15144482
125.0 1 0.15144482 0.39894623 0.15144482
125.0 2 0.15144482 0.36905858 0.15144482
125.0 3 0.15144482 0.330976 0.15144482
125.0 4 0.15144482 0.43948132 0.15144482
62.5 0 0.15144482 0.3747885 0.15144482
62.5 1 0.15144482 0.41568568 0.15144482
62.5 2 0.15144482 0.335385 0.15144482
62.5 3 0.15144482 0.24236794 0.15144482
62.5 4 0.15144482 0.307

0.15144482

In [34]:
accuracy(model)

0.5333333333333333

In [37]:
# run 3
initial = build_model()
T = 5.0
alpha = 0.8
T_min = 1
iters = 10

model,cost  = simulated_annealing(initial=initial,generate_f=update_NN, energy_f=energy_fun,T = T, min_T = T_min, iters = iters, alpha = alpha)

cost

5.0 0 0.22501716 0.42033005 0.22501716
5.0 1 0.22501716 0.36543468 0.22501716
5.0 2 0.21971859 0.21971859 0.22501716
5.0 3 0.21971859 0.29321182 0.21971859
5.0 4 0.21971859 0.3507562 0.21971859
5.0 5 0.08423306 0.08423306 0.21971859
5.0 6 0.08423306 0.42129165 0.08423306
5.0 7 0.08423306 0.42522514 0.08423306
5.0 8 0.08423306 0.57941854 0.08423306
5.0 9 0.08423306 0.37934792 0.08423306
4.0 0 0.08423306 0.45700046 0.37934792
4.0 1 0.08423306 0.41588816 0.37934792
4.0 2 0.08423306 0.34075528 0.37934792
4.0 3 0.08423306 0.40152472 0.34075528
4.0 4 0.08423306 0.33312684 0.34075528
4.0 5 0.08423306 0.39693433 0.33312684
4.0 6 0.08423306 0.27930564 0.33312684
4.0 7 0.08423306 0.45494097 0.27930564
4.0 8 0.08423306 0.3265855 0.27930564
4.0 9 0.08423306 0.34509256 0.27930564
3.2 0 0.08423306 0.5665856 0.27930564
3.2 1 0.08423306 0.41481322 0.27930564
3.2 2 0.08423306 0.2960132 0.27930564
3.2 3 0.08423306 0.22399732 0.27930564
3.2 4 0.08423306 0.32493538 0.22399732
3.2 5 0.08423306 0.3276114 0.

0.08423306

In [38]:
accuracy(model)

0.9333333333333333

#REPORT

##Result of backpropagation

ACC = 100 %
Time - couple of seconds.



## Result of "SA".

### run 1 :

T = 5

T_min = 0.1

alpha = 0.15

cost = 0.19

acc = 50%

Time - couple of minutes

### run 2 :
T = 1000.0

alpha = 0.5

T_min = 10

iters = 5

acc = 53%

Time - couple of minutes

### run 3 :
T = 5

alpha = 0.8

T_min = 1

iters = 10

acc = 93%

Time - couple of minutes


## Result
It is easy can be seen, that backpropagation is more suitable for the task of learning NN. However, we can see that SA also could be use and show pretty good result. However takes much more time and very unstable. SA is also more powerful tool and can solve many different tasks of optimization.